wa/instruments/perf.py

```diff
@@ -13,126 +13,385 @@
+                  description="""
+                  Dictionary of commands to be run after the workloads run
+                  (same format as ``commands``).
+                 """),
     ]
 
     def __init__(self, target, **kwargs):
         super(PerfInstrument, self).__init__(target, **kwargs)
         self.collector = None
 
     def initialize(self, context):
+        # pylint: disable=unused-argument
+        # pylint: disable=access-member-before-definition
+        # pylint: disable=attribute-defined-outside-init
+        if self.pre_commands is None:
+            self.pre_commands = PerfCommandDict({})
+        if self.post_commands is None:
+            self.post_commands = PerfCommandDict({})
+        if self.commands is None:
+            if self.optionstring is None:
+                self.optionstring = DEFAULT_OPTIONSTRING
+
+            if self.events is None:
+                self.events = DEFAULT_EVENTS
+
+            if isinstance(self.optionstring, str):
+                self.optionstring = [self.optionstring]
+
+            if not self.labels:
+                self.labels = ['default{}'.format(i)
+                               for i, _ in enumerate(self.optionstring)]
+            elif isinstance(self.labels, str):
+                self.labels = [self.labels]
+
+            if len(self.labels) != len(self.optionstring):
+                raise ValueError('Lengths of labels and optionstring differ')
+
+            self.commands = PerfCommandDict({
+                label: {
+                    'command': 'stat',
+                    'kwflags': {'event': self.events},
+                    'options': shlex.split(options),
+                    'args': {
+                        'command': 'sleep',
+                        'args': 1000,
+                    },
+                    'stderr': '&1',
+                    'stdout': 'stat.out',
+                }
+                for label, options in zip(self.labels, self.optionstring)
+            })
+        else:
+            for name in ['optionstring', 'events', 'labels']:
+                if self.__dict__[name] is not None:
+                    raise ValueError(
+                        '{} should not be passed if commands is'.format(name))
+
         self.collector = PerfCollector(self.target,
-                                       self.events,
-                                       self.optionstring,
-                                       self.labels,
-                                       self.force_install)
+                                       self.force_install,
+                                       self.pre_commands,
+                                       self.commands,
+                                       self.post_commands)
 
     def setup(self, context):
         self.collector.reset()
+        version = self.collector.execute('--version').strip()
+        context.update_metadata('versions', self.name, version)
```

Comment:
I hope this is being logged in the right "metadata".

I've noticed the following (optional) idea that hasn't been addressed (from the original post):

> The force_install flag is kept from the previous implementation. However, what is the "standard WA way" of providing the perf binary? We could write the path in the agenda (e.g. for comparing versions of perf), we could have a collection of perf binaries in a standard location and pick the ideal one (e.g. based on kernel version of the target) from within the WA perf instrument, we could (but this requires automating the automation, WA, and feels messy) have WA get the binary from a standard location on the host and overwrite that one between runs... Ideally, it would be useful to have both control of which binaries are being used and automation for finding the optimal version (as I believe the perf-kernel interface evolves with the kernel).

Based on ARM-software/devlib#395 and ARM-software/devlib#396, this may be useful but might be complicated to implement(?)

Reply:
Yes, this is the correct metadata.

WA has  a standard mechanism for "resource resolution" that allows providing alternative versions of resources via various means.

https://workload-automation.readthedocs.io/en/latest/developer_information.html#dynamic-resource-resolution

See how `dhrystone` workload obtains its executable for example:

https://github.com/ARM-software/workload-automation/blob/master/wa/workloads/dhrystone/__init__.py#L78

Reply:
But `dhrystone` is picking its binaries based on the ABI which is pretty easy to check with an ELF reader; what I was considering was to get the version of `perf` which (AFAIK) can only be obtained by running the binaries themselves ... which requires something able to run them. Another approach would require storing this information in the file system (_e.g._ in the name).

Let's drop this idea due to its complexity.

Reply:
>But dhrystone is picking its binaries based on the ABI which is pretty easy to check with an ELF reader;

Actually, it's picking the binary based on an arbitrary string, which happens to be the ABI -- we're not checking the ELF header, just resolving based on directory structure. So the same can be done for perf (it's basically what you're saying about encoding in the name, except we use directories rather than modifying the file name).

Reply:
I see, my bad for not looking further through the call stack of the code you shared!

I believe this is a good idea. However, this feature isn't required for using `perf` as a single version is enough in almost all cases. As users of WA-`perf` seem to have been satisfied with the single version approach until now, let's keep this feature for a potential future PR, once we are certain it will be necessary and used!